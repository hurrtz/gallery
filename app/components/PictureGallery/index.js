import React from 'react';
import PropTypes from 'prop-types';
import Card from '@material-ui/core/Card';
import CardMedia from '@material-ui/core/CardMedia';
import CardActions from '@material-ui/core/CardActions';
import IconButton from '@material-ui/core/IconButton';
import ZoomInIcon from '@material-ui/icons/ZoomIn';
import { makeStyles } from '@material-ui/core/styles';
import useMediaQuery from '@material-ui/core/useMediaQuery';

import { CAR_IMAGE_FILE_EXTENSION } from 'containers/GalleryPage/constants';

const useStyles = makeStyles(theme => ({
  card: {
    maxWidth: 345,
    marginBottom: '2em',
    flex: '0 0 auto',
    [theme.breakpoints.down('sm')]: {
      flex: '1 0 auto',
    },
  },
}));

const generateURL = (url, thumbnail) =>
  `https://${url}_${thumbnail ? '2' : '27'}.${CAR_IMAGE_FILE_EXTENSION}`;

const PictureGallery = ({ title, imageURLs, setImageURL, onShowFullSize }) => {
  const classes = useStyles();
  const handleShowFullImage = url => {
    setImageURL(generateURL(url, false));
    onShowFullSize(true);
  };
  const isMobile = useMediaQuery('(max-width: 600px)');

  const createMobileContent = url => (
    <Card
      key={url}
      className={classes.card}
      onClick={() => handleShowFullImage(url)}
      data-testid="card"
    >
      <CardMedia image={generateURL(url, true)} component="img" title={title} />
    </Card>
  );

  const createDesktopContent = url => (
    <Card key={url} className={classes.card}>
      <CardMedia image={generateURL(url, true)} component="img" title={title} />

      <CardActions>
        <IconButton
          aria-label="show image"
          onClick={() => handleShowFullImage(url)}
          data-testid="button"
        >
          <ZoomInIcon />
        </IconButton>
      </CardActions>
    </Card>
  );

  if (isMobile) {
    return imageURLs.map(url => createMobileContent(url));
  }

  return imageURLs.map(url => createDesktopContent(url));
};

PictureGallery.propTypes = {
  title: PropTypes.string.isRequired,
  imageURLs: PropTypes.arrayOf(PropTypes.string),
  onShowFullSize: PropTypes.func,
  setImageURL: PropTypes.func,
};

export default PictureGallery;
