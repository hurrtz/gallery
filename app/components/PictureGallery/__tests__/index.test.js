import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';

import PictureGallery from '../index';

function createMatchMedia(isMobile) {
  return () => ({
    matches: isMobile,
    addListener: () => {},
    removeListener: () => {},
  });
}

const enforceMobile = () => {
  window.matchMedia = createMatchMedia(true);
};

const enforceDesktop = () => {
  window.matchMedia = createMatchMedia(false);
};

const defaultProps = {
  imageURLs: [],
  title: 'Test',
};

const propsWithURLs = {
  ...defaultProps,
  imageURLs: ['imageA', 'imageB', 'imageC'],
};

describe('<PictureGallery />', () => {
  describe('And has mobile view', () => {
    beforeEach(() => {
      enforceMobile();
    });

    it('should render and match the snapshot', () => {
      const tree = renderer
        .create(<PictureGallery {...defaultProps} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render with given urls and match the snapshot', () => {
      const tree = renderer
        .create(<PictureGallery {...propsWithURLs} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should call related callback functions when the card is clicked', () => {
      const setImageURL = jest.fn();
      const onShowFullSize = jest.fn();
      const props = { ...propsWithURLs, setImageURL, onShowFullSize };

      const { getAllByTestId } = render(<PictureGallery {...props} />);

      fireEvent.click(getAllByTestId('card')[0]);

      expect(setImageURL).toHaveBeenCalledTimes(1);
      expect(onShowFullSize).toHaveBeenCalledTimes(1);
    });
  });

  describe('And has desktop view', () => {
    beforeEach(() => {
      enforceDesktop();
    });

    it('should render and match the snapshot', () => {
      const tree = renderer
        .create(<PictureGallery {...defaultProps} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render with given urls and match the snapshot', () => {
      const tree = renderer
        .create(<PictureGallery {...propsWithURLs} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should call related callback functions when the card is clicked', () => {
      const setImageURL = jest.fn();
      const onShowFullSize = jest.fn();
      const props = { ...propsWithURLs, setImageURL, onShowFullSize };

      const { getAllByTestId } = render(<PictureGallery {...props} />);

      fireEvent.click(getAllByTestId('button')[0]);

      expect(setImageURL).toHaveBeenCalledTimes(1);
      expect(onShowFullSize).toHaveBeenCalledTimes(1);
    });
  });
});
