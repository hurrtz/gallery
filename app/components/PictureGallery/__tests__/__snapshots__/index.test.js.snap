// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PictureGallery /> And has desktop view should render and match the snapshot 1`] = `null`;

exports[`<PictureGallery /> And has desktop view should render with given urls and match the snapshot 1`] = `
Array [
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
  >
    <img
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
      src="https://imageA_2.jpg"
      title="Test"
    />
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        aria-label="show image"
        className="MuiButtonBase-root MuiIconButton-root"
        data-testid="button"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
            <path
              d="M12 10h-2v2H9v-2H7V9h2V7h1v2h2v1z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>,
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
  >
    <img
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
      src="https://imageB_2.jpg"
      title="Test"
    />
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        aria-label="show image"
        className="MuiButtonBase-root MuiIconButton-root"
        data-testid="button"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
            <path
              d="M12 10h-2v2H9v-2H7V9h2V7h1v2h2v1z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>,
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
  >
    <img
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
      src="https://imageC_2.jpg"
      title="Test"
    />
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        aria-label="show image"
        className="MuiButtonBase-root MuiIconButton-root"
        data-testid="button"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
            <path
              d="M12 10h-2v2H9v-2H7V9h2V7h1v2h2v1z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>,
]
`;

exports[`<PictureGallery /> And has mobile view should render and match the snapshot 1`] = `null`;

exports[`<PictureGallery /> And has mobile view should render with given urls and match the snapshot 1`] = `
Array [
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
    data-testid="card"
    onClick={[Function]}
  >
    <img
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
      src="https://imageA_2.jpg"
      title="Test"
    />
  </div>,
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
    data-testid="card"
    onClick={[Function]}
  >
    <img
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
      src="https://imageB_2.jpg"
      title="Test"
    />
  </div>,
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1 MuiPaper-rounded"
    data-testid="card"
    onClick={[Function]}
  >
    <img
      className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
      src="https://imageC_2.jpg"
      title="Test"
    />
  </div>,
]
`;
